src/core/listener.c

```diff
@@ -75,16 +75,34 @@ MsQuicListenerOpen(
     QuicSiloAddRef(Listener->Silo);
 #endif
 
+    BOOLEAN RegistrationShuttingDown;
+
     BOOLEAN Result = CxPlatRundownAcquire(&Registration->Rundown);
     CXPLAT_DBG_ASSERT(Result); UNREFERENCED_PARAMETER(Result);
 
-    QuicTraceEvent(
-        ListenerCreated,
-        "[list][%p] Created, Registration=%p",
-        Listener,
-        Listener->Registration);
-    *NewListener = (HQUIC)Listener;
-    Status = QUIC_STATUS_SUCCESS;
+    CxPlatLockAcquire(&Registration->ListenerLock);
+    RegistrationShuttingDown = Registration->ShuttingDown;
+    if (!RegistrationShuttingDown) {
+        CxPlatListInsertTail(&Registration->Listeners, &Listener->RegistrationLink);
+    }
+    CxPlatLockRelease(&Registration->ListenerLock);
+
+    if (RegistrationShuttingDown) {
+        CxPlatRundownRelease(&Registration->Rundown);
+        CxPlatEventUninitialize(Listener->StopEvent);
+        CXPLAT_FREE(Listener, QUIC_POOL_LISTENER);
+        Listener = NULL;
+        Status = QUIC_STATUS_INVALID_STATE;
+        goto Error;
+    } else {
+        QuicTraceEvent(
+            ListenerCreated,
+            "[list][%p] Created, Registration=%p",
+            Listener,
+            Listener->Registration);
+        *NewListener = (HQUIC)Listener;
+        Status = QUIC_STATUS_SUCCESS;
+    }
```

Comment:
```suggestion
    }

    QuicTraceEvent(
        ListenerCreated,
        "[list][%p] Created, Registration=%p",
        Listener,
        Listener->Registration);
    *NewListener = (HQUIC)Listener;
    Status = QUIC_STATUS_SUCCESS;
```
No need to have the `else`, because the `if` has that `goto`.